### ISSUE TITLE
Mismatch between parquet statistics and parts silently drops data

### ISSUE THREAD

**Issue by j-r77** [dask/dask#8399]
**What happened**:
Hi,

I've had some difficulty reading in (older) parquet data files, due to not Dask not finding statistics for about 2/3 of the files.

I'm running the following simple code:
```python
df = dask.dataframe.read_parquet(urlpath, engine='fastparquet').persist()
```

Where urlpath is a glob path that resolves to 5 folders, each with a _metadata and a _common_metadata file, and 17 parquet files (one of the folders has 16).
The total number of partitions should be (17*4)+16 = 84, but df.npartitions returns 36; the other files are not included in the df.
After a bunch of trying to trace the problem through the code, I ended up putting a print here:
https://github.com/dask/dask/blob/7446308083e2f284229a11e9b3e330aa9e73cf96/dask/dataframe/io/parquet/core.py#L347
Which gives me:
```python
len(meta)=0, len(statistics)=36, len(parts)=84, len(index)=15
```

Due to the zip() on this line:
https://github.com/dask/dask/blob/7446308083e2f284229a11e9b3e330aa9e73cf96/dask/dataframe/io/parquet/core.py#L1143-L1150
Only the first 36 of the parts will be included in df.

**What you expected to happen**:

It is quite probable that the root problem here occurred back when this data was written (although I have no clue how to investigate that further).
I can also circumvent the problem by passing in gather_statistics=False.
However, I can't see how cases where len(statistics) != len(parts) would return reliable results, and would at least expect a warning here.
Does anyone have pointers on how to further investigate and/or improve Dask's handling of this (exceptional?) case?
Or an explanation on why the current behavior is desired?

**Minimal Complete Verifiable Example**:

Not possible without the data in question.

**Anything else we need to know?**:

/

**Environment**:

- Dask version: 2021.11.1
- Fastparquet version: 0.7.1 (currently, for reading, unclear which version was used for writing)
- Python version: 3.9.7
- Operating System: Ubuntu 20.04.3 LTS
- Install method (conda, pip, source): Conda

**Comment by jsignell** [comment]
Thanks for writing this up @j-r77 if you are able to create an example where all the data is generated, that would definitely help towards a speedy resolution. 

But I'll ping @martindurant and @rjzamora anyways in case they have ideas of how to proceed.

**Comment by martindurant** [comment]
At a quick skim of the report, I imagine indeed we don't test for a case where only some of the data contains statistics. The statistics should be aligned with the "parts", so that you only exclude parts when you can be sure there is no good data there.

**Comment by rjzamora** [comment]
>Where urlpath is a glob path that resolves to 5 folders, each with a _metadata and a _common_metadata file, and 17 parquet files (one of the folders has 16).

I suspect that passing in a glob with multiple directories is causing the issue here. I am fairly confident that the glob-pattern feature was designed to handle a collection of files (not a collection of datasets).

>Not possible without the data in question.

Hopefully this is an exaggeration, or the problem will likely never get fixed ;)  I will try to reproduce the problem with a toy directory-glob example (I have alrady confirmed that leaving statistics out of some files in a single directory does not lead to problems).

**Comment by rjzamora** [comment]
#8416 doesn't close this, but it should result in a clear UserWarning when the statistics and partitions are misaligned (and automatically bail on statistics processing).

I have not been able to reproduce a real case where misalignment occurs, but I will be happy to work on a fix for the root cause of this issue once a reproducer becomes available.

**Comment by j-r77** [comment]
Thanks for the update @rjzamora . The data in this case comes from one dataset, but is manually partitioned into the 5 separate folders. Columns are the same in every folder, and the glob works for our situation in most cases. These problems were found in older data that we are trying to reprocess.

The _metadata_ files are a black box for me, but I'd like to understand more about what's going on. Is there any way I can check them for peculiarities? For instance, you tested 'leaving statistics out of some files', but I have no clue how you would do that.

**Comment by rjzamora** [comment]
>Columns are the same in every folder, and the glob works for our situation in most cases

Thanks for the info! Can you clarify the glob pattern you are using? Is it just something like `"root_dir/*/*"`?  Also, are there multiple `_metadata` files, or just one within the root directory?  I ask about the `_metadata` file count, because there can never be more than one "global" metadata file  for proper `read_parquet` behavior.  If Dask detects the existence of any one of the `_metadata` files, it will ignore the others, and will assume that this single file contains all footer metadata for the entire dataset.

>The metadata files are a black box for me, but I'd like to understand more about what's going on. Is there any way I can check them for peculiarities? For instance, you tested 'leaving statistics out of some files', but I have no clue how you would do that.

You can always inspect the `_metadata` file with `fastparquet` or `pyarrow`. For example:

```python
import fastparquet
# Use _metadata file to generate a `ParquetFile` object
pf = fastparquet.ParquetFile("your-dataset/_metadata")
# Extract a list of row-group information
row_groups = pf.row_groups
# For each row_group, you can get a list of column-chunk information
column_chunks = row_groups[0].columns
# The column-chunk statistics are stored in `meta_data.statistics`
# for each element in `column_chunks`
print(column_chunks[0].meta_data.statistics)
```
**Example output**:
```
<class 'fastparquet.parquet_thrift.parquet.ttypes.Statistics'>
distinct_count: None
max: b'\x03\x00\x00\x00\x00\x00\x00\x00'
max_value: None
min: b'\x03\x00\x00\x00\x00\x00\x00\x00'
min_value: None
null_count: 0
```
If statistics were not written for this particular column chunk, the printed value would just be `None`.  I am actually not sure if `fastparquet` makes the statistics optional, but pyarrow does have [a write_statistics option for write_table](https://arrow.apache.org/docs/python/generated/pyarrow.parquet.write_table.html#pyarrow.parquet.write_table).

Regardless of these `_metadata`/statistics  questions, I would be interested to know if #8416 (just merged) avoids the error originally reported here?

**Comment by martindurant** [comment]
You might try a glob pattern like "root_dir/**.parquet" (or whatever the data files are called).

> I am actually not sure if fastparquet makes the statistics optional

Yes!
```
    stats: True|False|list(str)
        Whether to calculate and write summary statistics. If True (default), do it for
        every column; if False, never do; and if a list of str, do it only for those
        specified columns.
```

**Comment by j-r77** [comment]
Thanks for the fast responses all! Took me some time to test, but I can confirm that #8416 solves our problem.

With regards to the cause of the statistics mismatch @rjzamora :
I'm passing in a path of the form `parquet/year=2021/month=12/day=20/brandcode=*/*.parquet`
Every brandcode=x folder has its own `_metadata` and `_common_metadata` file, containing a dataset written separately but with the same schema as for the other brandcodes
So in this case, would Dask assume there is no metadata file available? Or take the first one it encounters?
In this concrete example, data from 2 days ago, Statistics are found for alphabetically the 1st, 4th, and 5th brand, while 2nd and 3rd are nowhere to be seen.
I've tried checking the `_metadata` files with fastparquet, but so far I've not found any differences between any of them, and all seem so have statistics for every file.

**Comment by martindurant** [comment]
Our code looks for a file named "_metadata", but it may well be that it's not getting the right directory root ("parquet/year=2021/month=12/day=20/"), where there is no such file in this case. fastparquet has a `root=` argument to `ParquetFile` that can be used for such tricky cases, I'm not immediately certain whether this is exposed in the dask call, or if pyarrow has similar.

**Comment by rjzamora** [comment]
When you specify `"parquet/year=2021/month=12/day=20/brandcode=*/*.parquet"`, dask will not find a `_metadata` file in the list of paths that is generated. Therefore, having multiple `_metadata` files is probably not the problem (although, Dask does assume that there can only be one of these files when it does find it).

**Comment by scharlottej13** [comment]
@j-r77 closing this issue since it seems your problem was resolved with #8416 and the advice from `@rjzamora` and `@martindurant`
